snapshot/volume-snapshotting-proposal.md

```diff
@@ -0,0 +1,509 @@
+
+* **Delete:**
+
+    1. The user deletes the `VolumeSnapshot`
+
+    2. The controller removes the on-disk snapshot. Note: snapshots have no notion of "reclaim policy" - there is
+       no way to recover the deleted snapshot.
+
+    3. The controller removes the `VolumeSnapshotData` object.
+
+* **Promote snapshot to PV:**
+
+    1. The user creates a persistent volume claim referencing the snapshot object in the annotation. The PVC must
+       belong to a `StorageClass` using the external volume snapshot provisioner.
+
+    2. The provisioner will use the `VolumeSnapshotData` object to create a persistent volume using the corresponding
+       volume snapshot plugin.
+
+    3. The PVC is bound to the newly created PV containing the data from the snapshot.
+
+
+The snapshot operation is a no-op for volume plugins that do not support snapshots via an API call (i.e. non-cloud storage).
+
+
+### API
+
+* The `VolumeSnapshot` object
+
+```
+// The volume snapshot object accessible to the user. Upon succesful creation of the actual
+// snapshot by the volume provider it is bound to the corresponding VolumeSnapshotData through
+// the VolumeSnapshotSpec
+type VolumeSnapshot struct {
+	metav1.TypeMeta `json:",inline"`
+	Metadata        metav1.ObjectMeta `json:"metadata"`
+
+	// Spec represents the desired state of the snapshot
+	// +optional
+	Spec VolumeSnapshotSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+
+	// Status represents the latest observer state of the snapshot
+	// +optional
+	Status VolumeSnapshotStatus `json:"status" protobuf:"bytes,3,opt,name=status"`
+}
+
+type VolumeSnapshotList struct {
+	metav1.TypeMeta `json:",inline"`
+	Metadata        metav1.ListMeta  `json:"metadata"`
+	Items           []VolumeSnapshot `json:"items"`
+}
+
+// The desired state of the volume snapshot
+type VolumeSnapshotSpec struct {
+	// PersistentVolumeClaimName is the name of the PVC being snapshotted
+	// +optional
+	PersistentVolumeClaimName string `json:"persistentVolumeClaimName" protobuf:"bytes,1,opt,name=persistentVolumeClaimName"`
+
+	// SnapshotDataName binds the VolumeSnapshot object with the VolumeSnapshotData
+	// +optional
+	SnapshotDataName string `json:"snapshotDataName" protobuf:"bytes,2,opt,name=snapshotDataName"`
+}
+
+type VolumeSnapshotStatus struct {
+	// The time the snapshot was successfully created
+	// +optional
+	CreationTimestamp metav1.Time `json:"creationTimestamp" protobuf:"bytes,1,opt,name=creationTimestamp"`
+
+	// Representes the lates available observations about the volume snapshot
+	Conditions []VolumeSnapshotCondition `json:"conditions" protobuf:"bytes,2,rep,name=conditions"`
+}
+
+type VolumeSnapshotConditionType string
+
+// These are valid conditions of a volume snapshot.
+const (
+	// VolumeSnapshotReady is added when the snapshot has been successfully created and is ready to be used.
+	VolumeSnapshotConditionReady VolumeSnapshotConditionType = "Ready"
```

Comment:
I am thinking to use the following conditions, CREATED, READY, FAILED with the following meanings

- CREATED(no) READY(no) FAILED(no): the snapshot request is pending or just initiated
- CREATED(yes) READY(no) FAILED(no): the snapshot request is received and issue, and the snapshot is cut (after this timepoint, user can resume their applications)
- CREATED(yes) READY(yes) FAILED(no): the snapshot is created and finishes uploading the data, it is ready to use.
- CREATED(yes/no) READY(no) FAILED(yes): the snapshot is not ready to use. It might failed before created or after during uploading.